### DISCUSSION TITLE
What is the difference between "bounds" and "x_length, y_length, z_length" in the cube pyvista method?

### DISCUSSION THREAD

**Discussion by juanccilleruelo** [pyvista/pyvista#3981] [Q&A]
If I use the following statement, the values "%_length" seem to be ignored. 

Cube2 = pv.Cube(bounds=(0.0, 1.5, 0.0, 1.0, 0.0, 1.5), x_length = 1, y_length = 1, z_length = 1.00) 
plotter.add_mesh(Cube2, opacity = 0.5, color = ''blue'') 
plotter.show_bounds(grid = ''front'', location = ''outer'', all_edges = True) 

On the other hand, I understand that bounds mean "dimensions" and not "limits." Is this true?

**Comment by banesullivan** [comment]
See https://docs.pyvista.org/api/utilities/_autosummary/pyvista.Cube.html#cube

> bounds : sequence, optional
>    Specify the bounding box of the cube. If given, all other size arguments are ignored. (xMin, xMax, yMin, yMax, zMin, zMax).

If bounds are given, then all other arguments specifying the extent will be ignored

> I understand that bounds mean "dimensions" and not "limits." Is this true?

Not sure what you mean by this... bounds are the spatial limits

**Reply by juanccilleruelo** [reply]
I rephrase the second question:

I understand that "bounds" in this context means "the size of the cube" and not the "limits where the cube is caged."

Because:

If I set bounds and the x_length, y_length, and z_length size arguments are ignored, his bounds define the size of the cube.  

Are Both asserts true?

**Reply by banesullivan** [reply]
The bounds are the "limits where the cube is caged": defining the spatial extent and by extension the size

**Reply by MatthewFlamm** [reply]
This is related to https://github.com/pyvista/pyvista/discussions/3960 for context.

**Reply by juanccilleruelo** [reply]
Can I specify the bounds and a different size for a cube?

For example, a (10, 10, 10) size and a (15, 15, 15) bounds.

**Reply by banesullivan** [reply]
There is no difference between size and bounds for the `Cube()`. What are you trying to accomplish? Do you want a grid with a resolution of `(10, 10, 10)` and bounds `(15, 15, 15)`?... If so, `UniformGrid` is what you are after

I'm lost on what you're after since there is no difference between size and bounds for a cube.

**Reply by juanccilleruelo** [reply]
In the source code, we can see this:

```
    if bounds is not None:
        if np.array(bounds).size != 6:
            raise TypeError(
                'Bounds must be given as length 6 tuple: (xMin, xMax, yMin, yMax, zMin, zMax)'
            )
        src.SetBounds(bounds)
    else:
        src.SetCenter(center)
        src.SetXLength(x_length)
        src.SetYLength(y_length)
        src.SetZLength(z_length)
```
        
As you said previously, you can see that the "x_length, y_length, and z_length" parameters are ignored.
But the cube is drawn, and the size of the cube is the space specified by the "bounds" parameter.

In this case, the "bounds" have the mean of "size."  Define the size of the cube, delimiting the space it occupies.

**Reply by juanccilleruelo** [reply]
> This is related to #3960 for context.

Yes, they are related because this question is an try to understand the other. 
But I want to keep two different questions distinct. 

This current question is about a very simple thing. Or not. I don't know. 😉

**Comment by luizenger** [comment]
> If I use the following statement, the values "%_length" seem to be ignored.
> 
> Cube2 = pv.Cube(bounds=(0.0, 1.5, 0.0, 1.0, 0.0, 1.5), x_length = 1, y_length = 1, z_length = 1.00) plotter.add_mesh(Cube2, opacity = 0.5, color = ''blue'') plotter.show_bounds(grid = ''front'', location = ''outer'', all_edges = True)
> 
> On the other hand, I understand that bounds mean "dimensions" and not "limits." Is this true?

When you pass the _bounds_ argument, you are specifically telling where the x_min / x_max / y_min / y_max / z_min / z_max of your cube are located in the reference frame. With _*_length_ and _center_ you are specifying the size of each side along that axis and where the center of your cube is to be located in the reference frame.

Let's say you want a cube with lengths 4 in all directions, placed in the origin of your system, the (0, 0, 0) position.
For this you can:
  - give bounds = (-2, 2, -2, 2, -2, 2) or
  - give x_length = 4, y_length = 4, z_length = 4, center = (0, 0, 0)

In both cases, the faces of the cube will be at the same position in your coordinate system.
By default, PyVista checks if you are passing the _bounds_ argument correctly, and if you are it then ignores the data for _*_length_ and _center_
